src/types.ts

```diff
@@ -509,16 +446,45 @@ export class NullableType<T extends AnyType> extends Type<Infer<T> | null> imple
   and<K extends AnyType>(schema: K): IntersectionType<this, K> {
     return new IntersectionType(this, schema);
   }
-  default(value: Nullable<Infer<T>> | (() => Nullable<Infer<T>>)) {
-    return new NullableType(this.schema, value);
+}
+
+// type X = {
+//   a: DefaultableType<StringType>;
+//   b: DefaultableType<NumberType>;
+// };
+
+type DS = DefaultableType<StringType>;
+
+type R = DS extends Type<any> ? (DS extends Type<infer K> ? K : DS) : any;
```

Comment:
I can't seem to be able to infer the type of of a DefaultableType<StringType> (substitute StringType for any other Type class)... It just goes to any... any insight?

Reply:
@pierreis

Reply:
I am not sure.
That is quite disturbing.

Let me check.

Reply:
I managed to get type Infer to worth with DefaultTypes by making Infer more exhaustive:

```typescript
export type Infer<T extends AnyType> = T extends DefaultableType<any>
  ? T extends DefaultableType<infer Schema>
    ? Schema extends Type<infer K>
      ? Eval<K>
      : never
    : never
  : T extends Type<infer K>
  ? Eval<K>
  : never;
```

But I still can't get the DefaultType<ObjectType<Shape>> to type the correct defaultValue (default keys as optional)

Reply:
Sounds good. Debugging advanced types in TypeScript is hellish...
I still have no clue.

Reply:
I've managed to get very close. However I now have the DeepPartialShape type become possibly infinitely deep, it keeps doing that sometimes when something gets more complicated in the type system. 

I find that typescript is extremely buggy once you go a little advanced... I'll keep at it for a while... But although we've managed to introduce DefaultableType I still can't type the keys of defaultValue of object as optional... Losing my mind haha...

Reply:
I am considering throwing away deepPartial... It has been a thorn in my side getting it to keep compiling

Reply:
I cannot really see a use-case from my point of view, so I wouldn't mind.

Reply:
I pushed with everything working except for the fact that deepPartialShape will no longer compile. Gonna sit on this for a little as I mull it over.